case_studies/dc2_cataloging/utils/image_normalizer.py

```diff
@@ -0,0 +1,242 @@
+
+    def num_channels_per_band(self):
+        pre_nch = super().num_channels_per_band()
+        return pre_nch + len(self.asinh_params["thresholds"])
+
+    def get_input_tensor(self, batch):
+        pre_input_tensor = super().get_input_tensor(batch)
+        raw_images = batch["images"][:, self.bands].unsqueeze(2)
+
+        filtered_images = raw_images - self.asinh_thresholds_tensor
+        processed_images = filtered_images * torch.exp(self.asinh_scales_tensor)
+        processed_images = torch.asinh(processed_images)
+
+        return (
+            torch.cat((pre_input_tensor, processed_images), dim=2)
+            if pre_input_tensor is not None
+            else processed_images,
+            torch.stack(
+                (self.asinh_thresholds_tensor.squeeze(), self.asinh_scales_tensor.squeeze())
+            ),
+        )
+
+
+class MovingAvgAsinhImageNormalizer(ImageNormalizer):
+    def __init__(
+        self,
+        bands: list,
+        include_original: bool,
+        include_background: bool,
+        concat_psf_params: bool,
+        num_psf_params: int,
+        log_transform_stdevs: list,
+        use_clahe: bool,
+        clahe_min_stdev: float,
+        asinh_params: Dict[str, float],
+    ):
+        super().__init__(
+            bands,
+            include_original,
+            include_background,
+            concat_psf_params,
+            num_psf_params,
+            log_transform_stdevs,
+            use_clahe,
+            clahe_min_stdev,
+        )
+        self.asinh_params = asinh_params
+
+        assert self.asinh_params, "asinh_params can't be None"
+        assert (
+            not self.include_background
+        ), "if you want to use asinh, please don't include background"
+
+        thresholds_num = len(self.asinh_params["thresholds"])
+        thresholds = torch.tensor(self.asinh_params["thresholds"]).view(1, 1, thresholds_num, 1, 1)
+        self.asinh_thresholds_tensor = thresholds
+
+        self.asinh_buffer_size = 1000
+        self.asinh_buffer_ptr = 0
+        self.asinh_quantiles_tensor = torch.tensor([0.01, 0.1, 0.25, 0.5, 0.75, 0.9, 0.99])
+        buffer_shape = (self.asinh_buffer_size, *self.asinh_thresholds_tensor.squeeze().shape)
+        self.asinh_thresholds_buffer = torch.full(buffer_shape, torch.nan)
+        # it's hard to get the device type of torch.nn.Module
+        # https://stackoverflow.com/questions/58926054/how-to-get-the-device-type-of-a-pytorch-module-conveniently
+        self.device_indicator = torch.nn.Parameter(torch.tensor([0]), requires_grad=False)
+
+    def num_channels_per_band(self):
+        pre_nch = super().num_channels_per_band()
+        return pre_nch + len(self.asinh_params["thresholds"])
+
+    def get_input_tensor(self, batch):
+        pre_input_tensor = super().get_input_tensor(batch)
+        raw_images = batch["images"][:, self.bands].unsqueeze(2)
+
+        self.asinh_thresholds_tensor = self.asinh_thresholds_tensor.to(
```

Comment:
I believe calling `to` is discouraged when you're using pytorch-lightning, because lightning is supposed to manage devices for you and calls to `to` prevent it from doing so.